src/liger_kernel/chunked_loss/orpo_loss.py

```diff
@@ -0,0 +1,211 @@
+        rejected_logps (torch.Tensor): Avg log probabilities of rejected tokens. Shape: (batch_size,).
+        beta (float): Weight for the odds ratio loss.
+    """
+    log_odds = (chosen_logps - rejected_logps) - (
+        torch.log1p(-torch.exp(chosen_logps)) - torch.log1p(-torch.exp(rejected_logps))
+    )
+    ratio = F.logsigmoid(log_odds)
+    return beta * ratio.sum()
+
+
+class LigerFusedLinearORPOFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        _input,
+        weight,
+        target,
+        bias=None,
+        ignore_index=-100,
+        beta=0.1,
+        compiled=True,
+    ):
+        """
+        Fused linear layer with ORPO (Odds-Ratio Preference Optimization) loss.
+        Handles both the forward and backward pass of the final linear layer with ORPO loss.
+        Inspired from LigerFusedLinearCrossEntropyFunction which fuses final linear layer and CE loss.
+
+        Args:
+            _input (torch.Tensor): Input tensor. Shape: (batch_size, seq_len, hidden_size).
+            weight (torch.Tensor): Weight tensor. Shape: (vocab_size, hidden_size).
+            target (torch.Tensor): Target tensor. Shape: (batch_size, seq_len).
+            bias (torch.Tensor, optional): Bias tensor. Shape: (vocab_size,).
+            ignore_index (int): Index to ignore for loss computation.
+            compiled (bool): Whether to use torch compile for chunk accumulation.
+        """
+        # TODO: Tune CHUNK_SIZE to fully utilize the GPU
+        CHUNK_SIZE = 1
+
+        def _compute_orpo_loss(input_chunk, weight, target_chunk, bias=None):
+            len_chosen_chunk = target_chunk.shape[0] // 2
+
+            unnorm_logits = input_chunk @ weight.t()  # chunk_size x V
+            if bias is not None:
+                unnorm_logits = unnorm_logits + bias
+            unnorm_logits = unnorm_logits.float()
+            norm_logits = F.log_softmax(unnorm_logits, dim=-1)
+
+            chosen_nll_loss = F.nll_loss(
+                norm_logits[:len_chosen_chunk].view(-1, norm_logits.shape[-1]),
+                target_chunk[:len_chosen_chunk].view(-1),
+                reduction="sum",
+                ignore_index=ignore_index,
+            )
+            chosen_nll_loss = (
+                chosen_nll_loss / (target[: target.shape[0] // 2] != ignore_index).sum()
+            )
+
+            loss_mask = target_chunk != ignore_index
+            label_chunk = torch.where(loss_mask, target_chunk, 0)
+
+            per_token_logps = norm_logits.gather(-1, label_chunk.unsqueeze(-1)).squeeze(
+                -1
+            )
+            average_log_prob = (per_token_logps * loss_mask).sum(-1) / loss_mask.sum(-1)
+
+            chosen_logps = average_log_prob[:len_chosen_chunk]
+            rejected_logps = average_log_prob[len_chosen_chunk:]
+
+            or_loss = odds_ratio_loss(chosen_logps, rejected_logps, beta=beta)
```

Comment:
could we rename this to odds ratio loss?

The returned is orpo loss = sft loss + odds ratio loss, not this one.

Reply:
To clarify, or_loss here is the odds ratio loss; while the loss returned from the fn is the sum of odds ratio loss (or_loss) and sft loss (chosen_nll_loss). Do you want to rename or_loss to something more verbose?

Reply:
Thanks for clarification. Could we rename or_loss to odds_ratio_loss since or_loss might be misunderstood to be orpo_loss.